lib/mixins/lazy-upgrade-mixin.html

```diff
@@ -0,0 +1,131 @@
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="../utils/boot.html">
+<link rel="import" href="../utils/mixin.html">
+<link rel="import" href="../utils/render-status.html">
+<script>
+  (function(){
+    'use strict';
+
+  /**
+   * Element class mixin that provides API for lazily upgrading Polymer Elements.
+   *
+   * This mixin can be used with the `disable-upgrade` attribute to lazily upgrade elements
+   * that are descendants of this element. This is useful for elements that need to display
+   * many elements.
+   *
+   * When upgrading elements, the mixin will upgrade as many as it can within the budget
+   * set by `_lazyUpgradeBudget`, which defaults to 16ms.
+   *
+   * When finished processing lazy-upgrade elements, this mixin will set `_lazyUpgrading`
+   * to false on the instance.
+   *
+   * @polymerMixin
+   * @memberof Polymer
+   * @summary Element class mixin that provides API for lazily upgrading Polymer elements
+   */
+    Polymer.LazyUpgradeMixin = Polymer.dedupingMixin(function(base) {
+
+      const LAZY_UPGRADE = 'lazy-upgrade';
+      const DISABLE_UPGRADE = 'disable-upgrade';
+      const LAZY_UPGRADE_QUERY = `[${LAZY_UPGRADE}]`;
+      const LAZY_UPGRADE_BUDGET = 16;
+
+      function sortCandidates(a, b) {
+        let orderA = parseInt(a.getAttribute(LAZY_UPGRADE), 10) || 0;
+        let orderB = parseInt(b.getAttribute(LAZY_UPGRADE), 10) || 0;
+        return orderA - orderB;
+      }
+
+      function findCandidates(root) {
+        let candidates = Array.from(root.querySelectorAll(LAZY_UPGRADE_QUERY));
+        candidates.sort(sortCandidates);
+        return candidates;
+      }
+
+      function upgradeNode(node) {
+        if (node.localName === 'dom-if') {
+          node.if = true;
+        }
+        node.removeAttribute(DISABLE_UPGRADE);
+        node.removeAttribute(LAZY_UPGRADE);
+      }
+
+      function eventInCurrentScope(scope, event) {
+        return event.target === event.composedPath()[0];
+      }
+
+      /** @polymerMixinClass */
+      return class LazyUpgrade extends base {
+        // add a configurable
+        static get properties() {
+          return {
+            /**
+             * Instance-level property for configuring the frame budget for lazy-upgrading elements.
+             * Defaults to 16ms
+             */
+            _lazyUpgradeBudget: {
+              type: Number,
+              value: LAZY_UPGRADE_BUDGET
+            },
+            /**
+             * Instance-level property that is shows when the element is lazy-upgrading elements
+             */
+            _lazyUpgrading: {
+              type: Boolean
+            }
+          }
+        }
+
+        ready() {
+          super.ready();
+          // check for dom-repeat and dom-if elements stamping lazy-upgrade nodes
+          this.addEventListener('dom-change', (ev) => {
```

Comment:
Also just curious... was there a use case for needing support for repeat/if templates stamping `lazy-upgrade` elements? Or is this just "just in case"?  I can maybe picture scenarios where it might be good, but seems cornery.